commandapi-annotations-reloaded/src/main/java/dev/jorel/commandapi/annotations/reloaded/parser/AnnotationParser.java

```diff
@@ -0,0 +1,46 @@
+/*******************************************************************************
+ * Copyright 2018, 2020 Jorel Ali (Skepter) - MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *******************************************************************************/
+package dev.jorel.commandapi.annotations.reloaded.parser;
+
+import javax.lang.model.element.Element;
+import java.util.Optional;
+
+/**
+ * Interface for parsing a CommandAPI annotation and retrieving some data.
+ *
+ * @param <AnElement>      The type of element that the annotation is applied to
+ * @param <AParserContext> The type of contextual information needed to parse the data
+ * @param <AReturnType>    The type of data returned
+ */
+public interface AnnotationParser<
+	AnElement extends Element,
+	AParserContext extends AnnotationParserContext<AnElement>,
+	AReturnType
+	> {
```

Comment:
Put this on one line

Reply:
Are you sure? It's going to make for some very long lines.

Reply:
Yeah, I know. If you are not too fond of this, at least put the closing angular bracket and the opening curly brace on the line above. Because this looks weird.

Reply:
There are some examples of a bunch of generics like this in other classes. For example, `AbstractArgumentTree`: 

https://github.com/JorelAli/CommandAPI/blob/aacd95712386f2342e051a0331306298645eec56/commandapi-core/src/main/java/dev/jorel/commandapi/AbstractArgumentTree.java#L14-L22

Honestly, a bit wonky too, so maybe there's a better style we want to establish consistently across the classes.

Also, iirc, the `@cond DOX` thing is around the generics like that to fix something with how Doxygen displays the javadocs (see https://github.com/JorelAli/CommandAPI/commit/e6f77968b97133d3739936abdd90fa58e1530b4e)? In this case, that would apply to the `AParserContext` type:

```java
AParserContext 
/// @cond DOX
extends AnnotationParserContext<AnElement>
/// @endcond
```

Reply:
I think it's best to use the same style as the rest of the project here then for consistency, and if we want to change it, we can have an issue to do that across the whole project.